src/ontology/solvers/backend/OntologyGraphSolvingStrategy.java

```diff
@@ -111,12 +113,16 @@ private void addPreferenceToCurBottom(ConstantSlot curBtm, Set<Constraint> consS
     }
 
     @Override
-    protected InferenceSolution mergeSolution(List<Map<Integer, AnnotationMirror>> inferenceSolutionMaps) {
+    protected InferenceResult mergeInferenceResults(List<Pair<Map<Integer, AnnotationMirror>, Collection<Constraint>>> inferenceResults) {
         Map<Integer, AnnotationMirror> result = new HashMap<> ();
         Map<Integer, EnumSet<OntologyValue>> ontologyResults = new HashMap<> ();
 
-        for (Map<Integer, AnnotationMirror> inferenceSolutionMap : inferenceSolutionMaps) {
-            for (Map.Entry<Integer, AnnotationMirror> entry : inferenceSolutionMap.entrySet()) {
+        for (Pair<Map<Integer, AnnotationMirror>, Collection<Constraint>> inferenceResult : inferenceResults) {
+            if (inferenceResult.fst == null) {
+                return new DefaultInferenceResult(inferenceResult.snd);
+            }
+
+            for (Map.Entry<Integer, AnnotationMirror> entry : inferenceResult.fst.entrySet()) {
```

Comment:
Can you put it into else branch to make it more clear?

Reply:
Importing `Map.Entry` and using `Entry` directly would be better.

Reply:
> Can you put it into else branch to make it more clear?

As there is a return statement in the `if` branch, it is obvious that line 125 only can be reached in the else branch. Adding an explicit else make the for loop further nested, and seems a bit not easy to read....

Reply:
Hmm, then it's ok to keep it.